certsuite/device_handler/adb_device_interface.py

```diff
@@ -0,0 +1,480 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import ConfigParser
+import datetime
+import os
+import sys
+import posixpath
+import re
+import shutil
+import tempfile
+import time
+import traceback
+import logging
+import subprocess
+
+import mozdevice
+import marionette
+import gaiautils
+import wptserve
+import moznetwork
+from mozlog.structured import get_default_logger
+from marionette_extension import AlreadyInstalledException
+from marionette_extension import install as marionette_install
+
+
+here = os.path.split(__file__)[0]
+
+class ADBDeviceInterface(object):
+
+    def __init__(self, logger, version):
+        try:
+            self.logger = logger
+            self.logger.info("Testing ADB connection")
+            self.device = ADBB2G()
+            self.version = version
+        except (mozdevice.ADBError, mozdevice.ADBTimeoutError) as e:
+            self.logger.critical('Error connecting to device via adb (error: %s). Please be ' \
+                            'sure device is connected and "remote debugging" is enabled.' % \
+                            e.msg)
+            return None
+
+    """
+    Please implement the following functions if your devices is not support android debug bridge
+    """
+    # Check the device is ready for executing testing
+    def check_device_is_ready(self):
+        return self.check_network()
+
+    # Check the device environment setting is good to go
+    def check_device_environment_setting(self):
+        return self.check_preconditions()
+
+
+    # Restart the device
+    def reboot(self):
+        self.restart_device()
+
+    def push_file(self, srcFile, desFile):
+        self.push_file(srcfile, desFile)
```

Comment:
This is a circular reference, and thus is meaningless.  Did you mean to do something like:

    def push_file(self, srcFile, desFile):
        return self.device.push_file(srcFile, desFile)